components/operations-controller/controllers/operation_controller.go

```diff
@@ -16,36 +16,423 @@ package controllers
+	}
+
+	ctx = tenant.SaveToContext(ctx, requestData.TenantID)
+	app, err := r.directorClient.FetchApplication(ctx, operation.Spec.ResourceID)
+	if err != nil {
+		logger.Error(err, operationError("Unable to fetch application", operation))
+
+		if isReconciliationTimeoutReached(operation, time.Duration(r.webhookConfig.TimeoutFactor)*r.webhookConfig.ReconciliationTimeout) {
+			if err := r.client.Delete(ctx, operation); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{RequeueAfter: r.webhookConfig.RequeueInterval}, nil
+	}
+
+	operation = setDefaultStatus(*operation)
+
+	if app.Result.Ready {
+		operation = setConditionStatus(*operation, v1alpha1.ConditionTypeReady, corev1.ConditionTrue, "")
+		operation.Status.Phase = v1alpha1.StateSuccess
+		if app.Result.Error != nil {
+			operation = setConditionStatus(*operation, v1alpha1.ConditionTypeError, corev1.ConditionTrue, *app.Result.Error)
+			operation.Status.Phase = v1alpha1.StateFailed
+		}
+		if err := r.client.Status().Update(ctx, operation); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if app.Result.Error != nil {
+		operation = setErrorStatus(*operation, *app.Result.Error)
+		operation.Status.Phase = v1alpha1.StateFailed
+		if err := r.client.Status().Update(ctx, operation); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	webhooks, err := retrieveWebhooks(app.Result.Webhooks, operation.Spec.WebhookIDs)
+	if err != nil {
+		logger.Error(err, operationError("Unable to retrieve webhooks", operation))
+		return ctrl.Result{}, nil
+	}
+
+	if isReconciliationTimeoutReached(operation, calculateReconciliationTimeout(webhooks, r.webhookConfig.WebhookTimeout)) {
+		opErr := errors.New("reconciliation timeout reached")
+		return r.updateStatusWithError(ctx, logger, operation, opErr)
+	}
+
+	webhookEntity := webhooks[operation.Spec.WebhookIDs[0]]
+	webhookStatus := &operation.Status.Webhooks[0]
+
+	if webhookStatus.WebhookPollURL == "" {
+		request := web_hook.NewRequest(webhookEntity, requestData, operation.Spec.CorrelationID)
+
+		response, err := r.webhookClient.Do(ctx, request)
+		if err != nil {
+			logger.Error(err, operationError("Unable to execute Webhook request", operation))
+			return r.handleWebhookTimeoutCheckResponse(ctx, logger, operation, webhookEntity, err)
+		}
+
+		switch *webhookEntity.Mode {
+		case graphql.WebhookModeAsync:
+			operation.Status.Webhooks[0].WebhookPollURL = *response.Location
+			if err := r.client.Status().Update(ctx, operation); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{Requeue: true}, nil
+		case graphql.WebhookModeSync:
+			return r.updateStatus(ctx, logger, operation)
+		default:
+			return ctrl.Result{}, errors.New("unsupported webhook mode")
+		}
+	}
+
+	requeueAfter, err := getNextPollTime(&webhookEntity, *webhookStatus, r.webhookConfig.TimeLayout)
+	if err != nil {
+		return ctrl.Result{}, err
```

Comment:
In case when parsing the requeue time shouldn't we set the operation status to Failed, as it will not be able to self-heal? At least the spec of the operation should be edited to fix that

Reply:
This will give more information that something bad has happened and we can further troubleshoot. Perhaps there is no need to send back information to director, but at least the Operation CRD can save that error, so we can later inspect it

Reply:
Yeah, I guess we should discuss that, as there are other similar situations like when the webhooks are not 1, or if the requestData fails to parse. All these are problems which should not happen and are sort of internal server errors which cannot self-heal on their own.

So I agree that here we should not return `err` as this would try to requeue. We've added an error log so that we can diagnose the issue, not sure if we should update the operation and/or notify director in such situations - like I said we have a bunch of those, so we should come to a unified agreement what to do in such situations.

Reply:
option 1 - we return the error and let controller manager do backoff requeues. We add some alerts about op being requeued more than X times. This is how we detect such bugs. Problem is that very often after we fix the bug, we wont be updating the op if the bug was related to how it was constructed or updated so the user will be patiently waiting for its reconcliation timeout.

option 2 - we just fail those directly, we still generate an alert - atleast this way the client won't be waiting for x hours for something that is doomed to fail :)

Reply:
I like Option 2 - which is related to what @pankrator  said in his second comment. Let's go with that one, as adequate alerts will take some time to prepare. With Option 2 the client will get feedback as soon as possible that something has gone wrong and not to wait for nothing.
Also, I think that we should do this same thing in the aforementioned scenarios as well, which are sort of internal server errors - request Data parsing fails, or webhooks > 1, etc.